gpkit/geometric_program.py

```diff
@@ -382,235 +101,182 @@ def _solve(self, solver=None, printing=True, skipfailures=False,
-
-    def _solve_sweep(self, printing, skipfailures, allownonoptimal):
-        """Runs a GeometricProgram through a sweep, solving at each grid point
+            result["cost"] = self.cost.subcmag(result["variables"])
+
+        result["sensitivities"] = {}
+        if "nu" in solver_out:
+            nu = np.array(solver_out["nu"]).ravel()
+            la = np.array([sum(nu[self.p_idxs == i])
+                           for i in range(len(self.posynomials))])
+        elif "la" in solver_out:
+            la = np.array(solver_out["la"]).ravel()
+            if len(la) == len(self.posynomials) - 1:
+                # assume the cost's sensitivity has been dropped
+                la = np.hstack(([1.0], la))
+            Ax = np.array(self.A.todense().dot(solver_out['primal'])).ravel()
+            z = Ax + np.log(self.cs)
+            m_iss = [self.p_idxs == i for i in range(len(la))]
+            nu = np.hstack([la[p_i]*np.exp(z[m_is])/sum(np.exp(z[m_is]))
+                            for p_i, m_is in enumerate(m_iss)])
+        else:
+            raise RuntimeWarning("The dual solution was not returned.")
+        result["sensitivities"]["monomials"] = nu
+        result["sensitivities"]["posynomials"] = la
+
+        self.result = result  # NOTE: SIDE EFFECTS AHOY
+
+        if solver_out.get("status", None) not in ["optimal", "OPTIMAL"]:
+            raise RuntimeWarning("final status of solver '%s' was '%s', "
+                                 "not 'optimal'." %
+                                 (solver, solver_out.get("status", None)) +
+                                 "\n\nTo generate a feasibility-finding"
+                                 " relaxed version of your\nGeometric Program,"
+                                 " run model.program.feasibility_search()."
+                                 "\n\nThe infeasible solve's result is stored"
+                                 " in the 'result' attribute\n(model.program.result)"
+                                 " and its raw output in 'solver_out'.")
+        else:
+            return result
+
+    def feasibility_search(self, varname=None, flavour="max", *args, **kwargs):
+        """Returns a new GP for the closest feasible point of the current GP.
 
         Arguments
         ---------
-        printing : bool (optional)
-            If True, then prints out sweep and GeometricProgram size.
+        varname : str
+            LaTeX name of slack variables.
+        flavour : str
+            Specifies the objective function minimized in the search:
 
-        Returns
-        -------
-        solution : dict
-            A dictionary containing the array of optimal values
-            for each free variable.
-        """
-        solution = GPSolutionArray()
-        solution.gp = self
+            "max" (default) : Apply the same slack to all constraints and
+                              minimize that slack. Described in Eqn. 10
+                              of [Boyd2007].
```

Comment:
How does this handle monomial equality constraints? As far as I can tell they are not known to `GeometricProgram`, so this must apply slack to each each side of the MEC as well?

That's a little odd -- we might eventually need a way to run `flavour="max"`, but only on a subset of the Constraints.